diff --git a/.gitignore b/.gitignore
index c01c56a..e05e9b6 100644
--- a/.gitignore
+++ b/.gitignore
@@ -9,4 +9,4 @@ DerivedData/
 /Package.resolved
 .ci/
 .docc-build/
-.swiftpm
+/bazel-*
diff --git a/BUILD.bazel b/BUILD.bazel
new file mode 100644
index 0000000..7c94a8a
--- /dev/null
+++ b/BUILD.bazel
@@ -0,0 +1,31 @@
+load("@rules_swift//swift:swift_library.bzl", "swift_library")
+
+# Copied from https://github.com/cgrindel/rules_swift_package_manager/tree/main
+
+load("@gazelle//:def.bzl", "gazelle", "gazelle_binary")
+
+# gazelle:exclude .build
+# gazelle:exclude Sources/swift-openapi-generator/Documentation.docc
+gazelle_binary(
+    name = "gazelle_bin",
+    languages = [
+        "@swift_gazelle_plugin//gazelle",
+    ],
+    visibility = ["//visibility:public"],
+)
+
+gazelle(
+    name = "create_build_files",
+    args = [
+        "-swift_dependency_index=$(execpath @swift_deps_info//:swift_deps_index)",
+    ],
+    data = ["@swift_deps_info//:swift_deps_index"],
+    gazelle = ":gazelle_bin",
+)
+
+swift_library(
+    name = "swift-openapi-generator-1.11",
+    srcs = ["Package@swift-6.0.swift"],
+    module_name = "swift-openapi-generator-1.11",
+    visibility = ["//visibility:public"],
+)
diff --git a/Package.resolved b/Package.resolved
new file mode 100644
index 0000000..a44fc3f
--- /dev/null
+++ b/Package.resolved
@@ -0,0 +1,78 @@
+{
+  "originHash" : "f0d630e371f43d01f72e61e6e20d32b6aa8362cb68dadd99f15cef17cf896f71",
+  "pins" : [
+    {
+      "identity" : "openapikit",
+      "kind" : "remoteSourceControl",
+      "location" : "https://github.com/mattpolzin/OpenAPIKit",
+      "state" : {
+        "revision" : "343b2c1793058fcc53c1bd7e2907f8e3a4d640fb",
+        "version" : "3.9.0"
+      }
+    },
+    {
+      "identity" : "swift-algorithms",
+      "kind" : "remoteSourceControl",
+      "location" : "https://github.com/apple/swift-algorithms",
+      "state" : {
+        "revision" : "87e50f483c54e6efd60e885f7f5aa946cee68023",
+        "version" : "1.2.1"
+      }
+    },
+    {
+      "identity" : "swift-argument-parser",
+      "kind" : "remoteSourceControl",
+      "location" : "https://github.com/apple/swift-argument-parser",
+      "state" : {
+        "revision" : "626b5b7b2f45e1b0b1c6f4a309296d1d21d7311b",
+        "version" : "1.7.1"
+      }
+    },
+    {
+      "identity" : "swift-collections",
+      "kind" : "remoteSourceControl",
+      "location" : "https://github.com/apple/swift-collections",
+      "state" : {
+        "revision" : "6675bc0ff86e61436e615df6fc5174e043e57924",
+        "version" : "1.4.1"
+      }
+    },
+    {
+      "identity" : "swift-http-types",
+      "kind" : "remoteSourceControl",
+      "location" : "https://github.com/apple/swift-http-types",
+      "state" : {
+        "revision" : "45eb0224913ea070ec4fba17291b9e7ecf4749ca",
+        "version" : "1.5.1"
+      }
+    },
+    {
+      "identity" : "swift-numerics",
+      "kind" : "remoteSourceControl",
+      "location" : "https://github.com/apple/swift-numerics.git",
+      "state" : {
+        "revision" : "0c0290ff6b24942dadb83a929ffaaa1481df04a2",
+        "version" : "1.1.1"
+      }
+    },
+    {
+      "identity" : "swift-openapi-runtime",
+      "kind" : "remoteSourceControl",
+      "location" : "https://github.com/apple/swift-openapi-runtime",
+      "state" : {
+        "revision" : "f039fa6d6338aab5164f3d1be16281524c9a8f89",
+        "version" : "1.11.0"
+      }
+    },
+    {
+      "identity" : "yams",
+      "kind" : "remoteSourceControl",
+      "location" : "https://github.com/jpsim/Yams",
+      "state" : {
+        "revision" : "deaf82e867fa2cbd3cd865978b079bfcf384ac28",
+        "version" : "6.2.1"
+      }
+    }
+  ],
+  "version" : 3
+}
diff --git a/Plugins/OpenAPIGenerator/BUILD.bazel b/Plugins/OpenAPIGenerator/BUILD.bazel
new file mode 100644
index 0000000..e64f346
--- /dev/null
+++ b/Plugins/OpenAPIGenerator/BUILD.bazel
@@ -0,0 +1,8 @@
+load("@rules_swift//swift:swift_binary.bzl", "swift_binary")
+
+swift_binary(
+    name = "OpenAPIGenerator",
+    srcs = ["plugin.swift"],
+    module_name = "OpenAPIGenerator",
+    visibility = ["//visibility:public"],
+)
diff --git a/Plugins/OpenAPIGeneratorCommand/BUILD.bazel b/Plugins/OpenAPIGeneratorCommand/BUILD.bazel
new file mode 100644
index 0000000..64015a7
--- /dev/null
+++ b/Plugins/OpenAPIGeneratorCommand/BUILD.bazel
@@ -0,0 +1,8 @@
+load("@rules_swift//swift:swift_binary.bzl", "swift_binary")
+
+swift_binary(
+    name = "OpenAPIGeneratorCommand",
+    srcs = ["plugin.swift"],
+    module_name = "OpenAPIGeneratorCommand",
+    visibility = ["//visibility:public"],
+)
diff --git a/Plugins/PluginsShared/BUILD.bazel b/Plugins/PluginsShared/BUILD.bazel
new file mode 100644
index 0000000..09f68d5
--- /dev/null
+++ b/Plugins/PluginsShared/BUILD.bazel
@@ -0,0 +1,13 @@
+load("@rules_swift//swift:swift_library.bzl", "swift_library")
+
+swift_library(
+    name = "PluginsShared",
+    srcs = [
+        "GeneratorMode.swift",
+        "PluginError.swift",
+        "PluginSource.swift",
+        "PluginUtils.swift",
+    ],
+    module_name = "PluginsShared",
+    visibility = ["//visibility:public"],
+)
diff --git a/Sources/PetstoreConsumerTestCore/BUILD.bazel b/Sources/PetstoreConsumerTestCore/BUILD.bazel
new file mode 100644
index 0000000..1c63290
--- /dev/null
+++ b/Sources/PetstoreConsumerTestCore/BUILD.bazel
@@ -0,0 +1,17 @@
+load("@rules_swift//swift:swift_library.bzl", "swift_library")
+
+swift_library(
+    name = "PetstoreConsumerTestCore",
+    srcs = [
+        "Assertions.swift",
+        "Common.swift",
+        "TestClientTransport.swift",
+        "TestServerTransport.swift",
+    ],
+    module_name = "PetstoreConsumerTestCore",
+    visibility = ["//visibility:public"],
+    deps = [
+        "@swiftpkg_swift_http_types//:HTTPTypes",
+        "@swiftpkg_swift_openapi_runtime//:OpenAPIRuntime",
+    ],
+)
diff --git a/Sources/_OpenAPIGeneratorCore/BUILD.bazel b/Sources/_OpenAPIGeneratorCore/BUILD.bazel
new file mode 100644
index 0000000..0a002b6
--- /dev/null
+++ b/Sources/_OpenAPIGeneratorCore/BUILD.bazel
@@ -0,0 +1,107 @@
+load("@rules_swift//swift:swift_library.bzl", "swift_library")
+
+swift_library(
+    name = "_OpenAPIGeneratorCore",
+    srcs = [
+        "Config.swift",
+        "Diagnostics.swift",
+        "DiagnosticsCollectorProvider.swift",
+        "Errors.swift",
+        "Extensions/Foundation.swift",
+        "Extensions/OpenAPIKit.swift",
+        "Extensions/String.swift",
+        "Extensions/SwiftStandardLibrary.swift",
+        "FeatureFlags.swift",
+        "GeneratorMode.swift",
+        "GeneratorPipeline.swift",
+        "GeneratorPipelineStage.swift",
+        "Hooks/DocumentFilter.swift",
+        "Hooks/FilteredDocument.swift",
+        "Layers/ParsedOpenAPIRepresentation.swift",
+        "Layers/RenderedSwiftRepresentation.swift",
+        "Layers/StructuredSwiftRepresentation.swift",
+        "Parser/ParserProtocol.swift",
+        "Parser/YamsParser.swift",
+        "Parser/validateDoc.swift",
+        "PlatformChecks.swift",
+        "Renderer/RendererProtocol.swift",
+        "Renderer/TextBasedRenderer.swift",
+        "Translator/ClientTranslator/ClientTranslator.swift",
+        "Translator/ClientTranslator/translateClientMethod.swift",
+        "Translator/CommonTranslations/translateAllAnyOneOf.swift",
+        "Translator/CommonTranslations/translateArray.swift",
+        "Translator/CommonTranslations/translateCodable.swift",
+        "Translator/CommonTranslations/translateObjectStruct.swift",
+        "Translator/CommonTranslations/translateRawEnum.swift",
+        "Translator/CommonTranslations/translateRawRepresentableEnum.swift",
+        "Translator/CommonTranslations/translateSchema.swift",
+        "Translator/CommonTranslations/translateStructBlueprint.swift",
+        "Translator/CommonTranslations/translateTypealias.swift",
+        "Translator/CommonTypes/Annotations.swift",
+        "Translator/CommonTypes/CommentExtensions.swift",
+        "Translator/CommonTypes/Constants.swift",
+        "Translator/CommonTypes/DiscriminatorExtensions.swift",
+        "Translator/CommonTypes/SchemaOverrides.swift",
+        "Translator/CommonTypes/StructBlueprint.swift",
+        "Translator/Content/CodingStrategy.swift",
+        "Translator/Content/ContentInspector.swift",
+        "Translator/Content/ContentType.swift",
+        "Translator/Content/SchemaContent.swift",
+        "Translator/FileTranslator.swift",
+        "Translator/FileTranslator+FeatureFlags.swift",
+        "Translator/Multipart/MultipartAdditionalProperties.swift",
+        "Translator/Multipart/MultipartContent.swift",
+        "Translator/Multipart/MultipartContentInspector.swift",
+        "Translator/Multipart/translateMultipart.swift",
+        "Translator/MultiplexTranslator.swift",
+        "Translator/Operations/OperationDescription.swift",
+        "Translator/Parameters/TypedParameter.swift",
+        "Translator/Parameters/translateParameter.swift",
+        "Translator/Recursion/DeclarationRecursionDetector.swift",
+        "Translator/Recursion/RecursionDetector.swift",
+        "Translator/RequestBody/TypedRequestBody.swift",
+        "Translator/RequestBody/translateRequestBody.swift",
+        "Translator/Responses/HTTPStatusCodes.swift",
+        "Translator/Responses/ResponseKind.swift",
+        "Translator/Responses/TypedResponse.swift",
+        "Translator/Responses/TypedResponseHeader.swift",
+        "Translator/Responses/acceptHeaderContentTypes.swift",
+        "Translator/Responses/translateResponse.swift",
+        "Translator/Responses/translateResponseHeader.swift",
+        "Translator/Responses/translateResponseOutcome.swift",
+        "Translator/ServerTranslator/ServerTranslator.swift",
+        "Translator/ServerTranslator/translateServerMethod.swift",
+        "Translator/TranslatorProtocol.swift",
+        "Translator/TypeAssignment/Builtins.swift",
+        "Translator/TypeAssignment/ReferenceStack.swift",
+        "Translator/TypeAssignment/SafeNameGenerator.swift",
+        "Translator/TypeAssignment/TypeAssigner.swift",
+        "Translator/TypeAssignment/TypeLocation.swift",
+        "Translator/TypeAssignment/TypeMatcher.swift",
+        "Translator/TypeAssignment/TypeName.swift",
+        "Translator/TypeAssignment/TypeUsage.swift",
+        "Translator/TypeAssignment/isSchemaSupported.swift",
+        "Translator/TypesTranslator/TypesFileTranslator.swift",
+        "Translator/TypesTranslator/translateAPIProtocol.swift",
+        "Translator/TypesTranslator/translateBoxedTypes.swift",
+        "Translator/TypesTranslator/translateComponentHeaders.swift",
+        "Translator/TypesTranslator/translateComponentParameters.swift",
+        "Translator/TypesTranslator/translateComponentRequestBodies.swift",
+        "Translator/TypesTranslator/translateComponentResponses.swift",
+        "Translator/TypesTranslator/translateComponents.swift",
+        "Translator/TypesTranslator/translateOperations.swift",
+        "Translator/TypesTranslator/translateSchemas.swift",
+        "Translator/TypesTranslator/translateServers.swift",
+        "Translator/TypesTranslator/translateServersVariables.swift",
+        "TypeOverrides.swift",
+        "YamlFileDiagnosticsCollector.swift",
+    ],
+    module_name = "_OpenAPIGeneratorCore",
+    visibility = ["//visibility:public"],
+    deps = [
+        "@swiftpkg_openapikit//:OpenAPIKitCompat",
+        "@swiftpkg_swift_algorithms//:Algorithms",
+        "@swiftpkg_swift_collections//:OrderedCollections",
+        "@swiftpkg_yams//:Yams",
+    ],
+)
diff --git a/Sources/swift-openapi-generator/BUILD.bazel b/Sources/swift-openapi-generator/BUILD.bazel
new file mode 100644
index 0000000..1f4ea12
--- /dev/null
+++ b/Sources/swift-openapi-generator/BUILD.bazel
@@ -0,0 +1,23 @@
+load("@rules_swift//swift:swift_binary.bzl", "swift_binary")
+
+swift_binary(
+    name = "swift-openapi-generator",
+    srcs = [
+        "Extensions.swift",
+        "FilterCommand.swift",
+        "GenerateCommand.swift",
+        "GenerateOptions.swift",
+        "GenerateOptions+runGenerator.swift",
+        "PluginSource.swift",
+        "Tool.swift",
+        "UserConfig.swift",
+        "runGenerator.swift",
+    ],
+    visibility = ["//visibility:public"],
+    deps = [
+        "//Sources/_OpenAPIGeneratorCore",
+        "@swiftpkg_openapikit//:OpenAPIKit",
+        "@swiftpkg_swift_argument_parser//:ArgumentParser",
+        "@swiftpkg_yams//:Yams",
+    ],
+)
